python/dgl/transform.py

```diff
@@ -673,71 +676,693 @@ def metapath_reachable_graph(g, metapath):
+            g = add_nodes(g, v_max - num_of_v, ntype=v_type)
+
+    num_nodes_dict = {}
+    for c_ntype in g.ntypes:
+        num_nodes_dict[c_ntype] = g.number_of_nodes(c_ntype)
+
+    graph_data = {}
+    for c_etype in g.canonical_etypes:
+        old_u, old_v = g.edges(form='uv', order='eid', etype=c_etype)
+        if c_etype == (u_type, e_type, v_type):
+            graph_data[c_etype] = (F.cat([old_u, u], dim=0),
+                                   F.cat([old_v, v], dim=0))
+        else:
+            graph_data[c_etype] = (old_u, old_v)
+
+    new_g = heterograph(graph_data,
+                        num_nodes_dict,
+                        idtype=g.idtype,
+                        device=g.device)
+
+    # copy node features
+    for c_ntype in g.ntypes:
+        for key, val in g.nodes[c_ntype].data.items():
+            new_g.nodes[c_ntype].data[key] = val
+
+    # copy edge features
+    for c_etype in g.canonical_etypes:
+        if c_etype == (u_type, e_type, v_type):
+            edata = g.edges[c_etype].data
+            # existing features
+            for key, val in edata.items():
+                if data is not None and data.get(key, None) is not None:
+                    new_feats = utils.prepare_tensor(new_g, data[key], 'data')
+                    assert len(new_feats) == len(u), \
+                        'Data length of {} should be {}, but got {}'.format(key,
+                                                                            len(u),
+                                                                            len(new_feats))
+                    new_g.edges[c_etype].data[key] = F.cat([val, new_feats], dim=0)
+                else:
+                    shape = F.shape(val)
+                    shape = (len(u),) + shape[1:]
+                    new_feats = F.zeros(shape, F.dtype(val), F.context(val))
+                    new_g.edges[c_etype].data[key] = F.cat([val, new_feats], dim=0)
+
+            # non-existing features
+            if data is not None:
+                for key, val in data.items():
+                    if edata.get(key, None) is None:
+                        shape = F.shape(data[key])
+                        shape = (g.number_of_edges(c_etype),) + shape[1:]
+                        new_feats = utils.prepare_tensor(new_g, data[key], 'data')
+                        assert len(new_feats) == len(u), \
+                            'Data length of {} should be {}, but got {}'.format(key,
+                                                                                len(u),
+                                                                                len(new_feats))
+                        new_g.edges[c_etype].data[key] = \
+                            F.cat([F.zeros(shape, F.dtype(val), F.context(val)),
+                                   new_feats], dim=0)
+        else:
+            for key, val in g.edges[c_etype].data.items():
+                new_g.edges[c_etype].data[key] = val
+
```

Comment:
Maybe a correct implementation for outplace mutation should be:
```python
def add_edges(g, u, v):
    g = g.clone()
    g.add_edges(u, v)
    return g
```
Especially if we will change some internal members in the future, this will always be safe.